Kotlin client: add volley library support

alisters opened the pull request:
<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->
Add a volley library to the kotllin client 

https://github.com/OpenAPITools/openapi-generator/issues/10011

I will add design notes to the above issue, but will move those notes to appropriate locations in the documentation when i know where they should reside. 

@jimschubert @wing328

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (5.3.0), `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

wing328 commented:
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/OpenAPITools/openapi-generator/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

wing328 commented:
cc @jimschubert (2017/09) ❤️, @dr4ke616 (2018/08) @karismann (2019/03) @Zomzog (2019/04) @andrewemery (2019/10) @4brunu (2019/11) @yutaka0m (2020/03)

wing328 commented:
@alisters can you please update the samples again as I saw some changes after pulling it for testing locally?

I ran `gradle build` but got the following errors:
```
$ gradle test

> Configure project :
WARNING: API 'variantOutput.getPackageLibrary()' is obsolete and has been replaced with 'variant.getPackageLibraryProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variantOutput.getPackageLibrary(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
WARNING: API 'variant.getJavaCompile()' is obsolete and has been replaced with 'variant.getJavaCompileProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variant.getJavaCompile(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':testReleaseUnitTest'.
> SDK location not found. Define location with an ANDROID_SDK_ROOT environment variable or by setting the sdk.dir path in your project's local properties file at '/Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/local.properties'.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/7.0.2/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 5s

```

Does the build work for you in your local environment?

4brunu commented:
This PR is very big, so it's difficult to review, but at least it doesn't affect the other generators.

alisters commented:
@wing328 
- The build does work locally for me. I should have called this out - you do need the android sdk installed locally though, and the ANDROID_SDK_ROOT environment variable defined. I'm not an expert in either android or gradle so i can't offer expert advice here - but volley itself pulls in android dependencies (https://github.com/google/volley/blob/master/build.gradle) so i think this is unavoidable ?  My output is below. 
- I'll update the samples
- We borrowed the mustache template for the gradle build from the existing android generator library. All suggestions for improvements there welcome !
=============
.\gradlew build
Starting a Gradle Daemon (subsequent builds will be faster)

> Configure project :
WARNING: API 'variantOutput.getPackageLibrary()' is obsolete and has been replaced with 'variant.getPackageLibraryProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variantOutput.getPackageLibrary(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.
WARNING: API 'variant.getJavaCompile()' is obsolete and has been replaced with 'variant.getJavaCompileProvider()'.
It will be removed in version 5.0 of the Android Gradle plugin.
For more information, see https://d.android.com/r/tools/task-configuration-avoidance.
To determine what is calling variant.getJavaCompile(), use -Pandroid.debug.obsoleteApi=true on the command line to display more information.

> Task :lint
Ran lint on variant debug: 9 issues found
Ran lint on variant release: 9 issues found
Wrote HTML report to file:///C:/dev/apps/gbs/api-central/clients/ib-integration-kotlin-volley/build/reports/lint-results.html
Wrote XML report to file:///C:/dev/apps/gbs/api-central/clients/ib-integration-kotlin-volley/build/reports/lint-results.xml

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.8.3/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 1m 29s
53 actionable tasks: 9 executed, 44 up-to-date

alisters commented on modules/openapi-generator/src/main/resources/kotlin-client/IStorable.mustache:
```diff
@@ -0,0 +1,9 @@
+{{>licenseInfo}}
```
@4brunu Good spot, it's definitely out of place. We had an issue with where to locate this file as it was a "supporting file" but we were trying to understand the system. I'll relook at it's location. 

As for the size of the PR, apologies if it's unwieldy - i wasn't sure how to approach this. We landed on trying to provide a largely functional library which made it quite large. We have tried to extend but not modify the system as much as possible in both code and templates to minimise the size and maintainability of the library. I'm guessing most PR's are more minor additions and full additional libraries or generators or less common. Maybe we can add to the contributors guidelines with some information for those adding a whole new library. e.g as i type this i wonder if a convenient way for you to build this would have been for me to provide you with a dockerfile that could set up the build environment. I'm having a look for any appropriate docker images now

alisters commented:
@wing328 - samples updated

alisters commented:
@wing328 et al. We are looking at exposing a bit more Volley functionality. e.g we've discovered locally we want to set the retry policy. This and other functionality is set on the request object. So there will be a minor refactor hopefully coming tomorrow that allows requests objects to be "post processed" - have other setter functions called on them after they're constructed. The header factories for auth will morph slightly into "post processors" as well

alisters commented:
@wing328 - this is ready for review now. "Ready" in a v0.1, easy does it on openapi beginners kind of way please 🙂  

Where is the best spot for doco ? Just in the generated readme thats embedded in the library ? 

Due to the way we've coded it, we can inject additional methods that supply headers and modify the request. This is mostly optional, but the two use cases people are likely to commonly use this in is injecting headers like trace headers, or using auth. I'd like to supply some good examples of this being done to help new users. e.g here is an example (excuse the formatting) of both setting this up for DI (using hilt in this case), and a post processor method that will call a setter on each request object - in this case setting the retry policy. 

> 

@Provides
    internal fun provideAddressApi(
        context: Context,
        restService: IRestService,
        configurationService: IConfigurationService,
        sessionService: ISessionService
    ): AddressApi {
        return AddressApi(
            context = context,
            requestQueue = restService.getRequestQueue(),
            requestFactory = RequestFactory(listOf(createSessionHeaderFactory(sessionService), createTraceHeaderFactory()), postProcessors = listOf(retryPolicySetter())),
            basePath = configurationService.getBaseUrl()
        )
    }

  private fun retryPolicySetter() : (Request<*>) -> Unit = { request: Request<*> -> Unit.apply { request.setRetryPolicy(
        DefaultRetryPolicy()
    ) } }

wing328 commented:
> Where is the best spot for doco ? Just in the generated readme thats embedded in the library ?

Yes, usually the auto-generated readme is the best place to document the usage of the auto-generated SDK.

wing328 commented:
After installing Android SDK and setting ANDROID_SDK_ROOT, I now get different errors:
```
> Task :compileDebugKotlin FAILED
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/apis/PetApi.kt: (176, 42): Unresolved reference: CSVParams
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/apis/PetApi.kt: (244, 38): Unresolved reference: CSVParams
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/apis/PetApi.kt: (445, 56): Unresolved reference: Field
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/apis/PetApi.kt: (445, 100): Unresolved reference: Field
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/apis/PetApi.kt: (513, 49): Unresolved reference: Part
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/apis/PetApi.kt: (513, 120): Unresolved reference: Part
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/apis/PetApi.kt: (513, 131): Unresolved reference: MultipartBody
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/BigDecimalAdapter.kt: (3, 12): Unresolved reference: squareup
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/BigDecimalAdapter.kt: (4, 12): Unresolved reference: squareup
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/BigDecimalAdapter.kt: (8, 6): Unresolved reference: ToJson
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/BigDecimalAdapter.kt: (13, 6): Unresolved reference: FromJson
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/BigIntegerAdapter.kt: (3, 12): Unresolved reference: squareup
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/BigIntegerAdapter.kt: (4, 12): Unresolved reference: squareup
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/BigIntegerAdapter.kt: (8, 6): Unresolved reference: ToJson
e: /Users/williamcheng/Code/openapi-generator/samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/BigIntegerAdapter.kt: (13, 6): Unresolved reference: FromJson

```

Do thees look familiar to you?

alisters commented:
@wing328 - sorry, this on Friday. They don't look familiar but i'll investigate

alisters commented:
@wing328 - i can reproduce these. I'll let you know when they are fixed - hopefully tomorrow. Apologies for the back and forth.

Is there a test harness the generated client is run against ? I haven't looked at the petstore yaml in a while, and i notice there are some comments around the authorization filter. I've only implemented basic auth, as a test for authorization, since we don't use the auth in open api for our own api's. We've also had no call for multipart form uploads so we haven't looked at implementing that. We made the design pretty extensible, but weren't looking at full feature coverage. 
Additionally i wondered whether there is a client feature matrix that is maintained at all - so i could indicate what we have and haven't done ?

alisters commented:
@wing328 - those errors should be resolved now. We hadn't run against petstore.yaml for a while - been using our own in house api's - sorry. I'll regenerate the samples shortly as well.

wing328 commented:
Can you please resolve the merge conflicts when you've time? I'll try to take another look this coming week.

wing328 commented:
I'll try to take another look later this week.

cc @jimschubert (2017/09) ❤️, @dr4ke616 (2018/08) @karismann (2019/03) @Zomzog (2019/04) @andrewemery (2019/10) @4brunu (2019/11) @yutaka0m (2020/03)

4brunu commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java:
```diff
@@ -183,6 +192,8 @@ public KotlinClientCodegen() {
         supportedLibraries.put(JVM_RETROFIT2, "Platform: Java Virtual Machine. HTTP client: Retrofit 2.6.2.");
         supportedLibraries.put(MULTIPLATFORM, "Platform: Kotlin multiplatform. HTTP client: Ktor 1.6.0. JSON processing: Kotlinx Serialization: 1.2.1.");
+        // TODO: Add version information.
+        supportedLibraries.put(JVM_VOLLEY, "Platform: JVM for Android. HTTP client: Volley. JSON processing: gson");
```
@alisters could you please fix this TODO?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,125 @@
+    testImplementation "org.robolectric:robolectric:${robolectric_version}"
+    testImplementation "net.jodah:concurrentunit:${concurrent_unit_version}"
+    {{#generateRoomModels}}
+    annotationProcessor "androidx.room:room-runtime:2.3.0-alpha03"
+    implementation "androidx.room:room-runtime:2.3.0-alpha03"
```
@alisters could you please update this to a stable release?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,125 @@
+}
+
+ext {
+    kotlin_version = '1.5.10'
```
@alisters could you please use this variable to keep `org.jetbrains.kotlin:kotlin-gradle-plugin` and `org.jetbrains.kotlin:kotlin-stdlib-jdk8` with the same version?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,125 @@
+
+dependencies {
+    coreLibraryDesugaring 'com.android.tools:desugar_jdk_libs:1.1.5'
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version"
```
To support Java 6 and 7, could you please use the `org.jetbrains.kotlin:kotlin-stdlib` instead? Or is there any dependency with `org.jetbrains.kotlin:kotlin-stdlib-jdk8`?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,125 @@
+
+android {
+    compileSdkVersion 29
+    buildToolsVersion '30.0.3'
```
@alisters `buildToolsVersion` is not needed anymore, could you please remove it?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,125 @@
+{{/useAndroidMavenGradlePlugin}}
+
+android {
+    compileSdkVersion 29
```
@alisters Currently the minimum sdk version required is 30, so could you please update this and algo `targetSdkVersion`?

alisters replied:
Very light in experience with Android. Does this look correct ?
compileSdkVersion 30
    defaultConfig {
        minSdkVersion 21
        targetSdkVersion 30
    }

4brunu replied:
Looks good to me 👍

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,125 @@
+allprojects {
+    repositories {
+        google()
+        jcenter()
```
@alisters could you please replace `jcenter` with `mavenCentral`?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,125 @@
+        maven {
+            url 'https://dl.google.com/dl/android/maven2'
+        }
+        jcenter()
```
@alisters could you please replace `jcenter` with `mavenCentral`?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/api.mustache:
```diff
@@ -0,0 +1,129 @@
+* Volley.newRequestQueue(context.applicationContext, myCustomHttpStack)
+*/
+class {{classname}} (
+    val context: Context,
+    val requestQueue: Lazy<RequestQueue> = lazy(initializer = {
+        Volley.newRequestQueue(context.applicationContext)
+    }),
+    val requestFactory: IRequestFactory = RequestFactory(),
+    val basePath: String = "{{{basePath}}}",
```
@alisters could you please make all of those private?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/api.mustache:
```diff
@@ -0,0 +1,129 @@
+        {{/required}}
+        {{/allParams}}
+
+        var contentTypes : Array<String> = arrayOf({{#consumes}}"{{{mediaType}}}"{{^-last}},{{/-last}}{{/consumes}})
+        var contentType: String = if (contentTypes.size > 0) { contentTypes[0] } else { "application/json" }
```
@alisters could you replace var with val?
Could you please update the if else to `if (contentTypes.isNotEmpty()) { contentTypes.first() } else { "application/json" }`?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/api.mustache:
```diff
@@ -0,0 +1,129 @@
+    {{/isDeprecated}}
+    suspend fun {{operationId}}({{^allParams}}){{/allParams}}{{#allParams}}{{>queryParams}}{{>pathParams}}{{>headerParams}}{{>bodyParams}}{{>formParams}}{{^-last}}, {{/-last}}{{#-last}}{{#operationAuthMethod}}, opAuthHeaderFactory = () -> map<String, String>{{/operationAuthMethod}}){{/-last}}{{/allParams}}: {{#returnType}}{{{returnType}}}?{{/returnType}}{{^returnType}}Unit{{/returnType}} {
+        {{#bodyParam}}
+        var body: Any? = {{paramName}}
+        {{/bodyParam}}
+        {{^bodyParam}}
+        var body: Any? = null
```
@alisters could you please make the body a val?

alisters replied:
Done

4brunu commented on samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/DateAdapter.kt:
```diff
@@ -0,0 +1,37 @@
+import java.util.Date
+import java.util.Locale
+
+class DateAdapter(val formatter: DateFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ", Locale.getDefault())) : TypeAdapter<Date>() {
```
@alisters this adapter is not used, could you please remove it?

alisters replied:
@4brunu - This was there already, and i think it might be used by other libraries that use gson like retrofit ? And also other serialisation libraries too (kotlinx). Looking in KotlinClientCodegen.java 

           case gson:
                supportingFiles.add(new SupportingFile("jvm-common/infrastructure/DateAdapter.kt.mustache", infrastructureFolder, "DateAdapter.kt"));

I could comment out or remove this supporting file for the gson serializer if you know it's not used for the other clients ? Otherwise i don't want to break other peoples clients, and i haven't codegen'd this generator for all different library/clients.

4brunu replied:
@alisters thanks for point that out, that is a bug, it's just old code that hasn't been removed.
This PR cleans up that.
https://github.com/OpenAPITools/openapi-generator/pull/11082

Could you please also remove DateAdapter and references to java.util.Date?
Thanks

alisters replied:
@4brunu - our file is in the samples, and generated code. Will your PR remove this for us ? Other than generating it, and including it as a supporting file i don't think we reference it. Sorry if i'm missing the point here

4brunu replied:
No, my PR only deletes that in the existing projects. Could you please remove this class and also the references to this?

alisters replied:
Done.

alisters replied:
@4brunu - i don't think i thought about this. This file is generated, and it's template file is common, in jvm-common/infrastructure/.  So i've removed the DateAdapter.mustache, but it now causes other libraries generated sample code to not compile. For the moment we have regenerated and pushed only _our_ samples - so that all samples still build. Assuming that you're PR will go in before ours, i can merge from master at that point and then all the libraries, ours included will no longer use DateAdapter. Does that sound ok ?

4brunu replied:
Yes, this is indeed a problem.
We have two options, merge first my PR, or I can merge my PR to your PR.

4brunu replied:
I just noticed now, that my PR was merged during the night, so this will no longer be an issue.
But maybe you will need to merge the OpenAPI master branch into your branch.

alisters replied:
Merged and repushed

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/request/GsonRequest.mustache:
```diff
@@ -0,0 +1,119 @@
+    errorListener: Response.ErrorListener
+) : Request<T>(method, url, errorListener) {
+
+    val gsonBuilder: GsonBuilder = GsonBuilder()
+        .registerTypeAdapter(OffsetDateTime::class.java, OffsetDateTimeAdapter())
+        .registerTypeAdapter(LocalDateTime::class.java, LocalDateTimeAdapter())
+        .registerTypeAdapter(LocalDate::class.java, LocalDateAdapter())
+        .registerTypeAdapter(ByteArray::class.java, ByteArrayAdapter())
+        .apply {
+            gsonAdapters?.forEach {
+                this.registerTypeAdapter(it.key, it.value)
+            }
+        }
+
+    val gson: Gson by lazy {
+        gsonBuilder.create()
+    }
```
@alisters why don't we use the Serializer.gson instead of creating a new Gson here?

alisters replied:
The consumer of the this, the builder of the gson request, needs to pass custom type adapters. And the types that respective builders know about are limited to that builder - so a singleton gson adapter wouldn't work. See code example below - In this case the payeeTypeAdapter is used to cope with sub typing. Thats the design rationale. That said, maybe there is a better way to do this and default to Serializer.gson if no gsonAdapters are provided ? I'm thinking about it :-)

#####
 @Provides
    internal fun providePayeeApi(
        context: Context,
        restService: IRestService,
        configurationService: IConfigurationService,
        sessionService: ISessionService
    ): PayeeApi {
        val payeeAdapter = mapOf(Pair<Type, Object>(Payee::class.java, PayeeTypeAdapter() as Object))
        return PayeeApi(
            context = context,
            requestQueue = restService.getRequestQueue(),
            requestFactory = RequestFactory(
                headerFactories = listOf(createSessionHeaderFactory(sessionService, restService), createTraceHeaderFactory()),
                postProcessors = listOf(retryPolicySetter),
                gsonAdapters = payeeAdapter
            ),
            basePath = configurationService.getBaseUrl()
        )
    }

4brunu replied:
Thanks for the explanation, some of my questions are because the volley integration is very different from the others in the kotlin generator

alisters replied:
@4brunu - general comments, no reply necessary. 

The other point i'd look at for these serialisers is the thread safety. I'm not sure if they are thread safe, and volley, especially with our non blocking co-routines, can definitely dispatch multiple calls in parallel. 

It is different, and TBH we couldn't decide which generator to use to start with, or whether to create a new one. To add a bit of the back story - In the related issue/feature request i mentioned there was 
"A volley client library in the android generator, which generates java.
A kotlin client generator with various libraries but not Volley.
Searching the issues shows the volley client may need attention - #9293 "

So the major reasons to put it in the Kotlin client 
- were to reuse the kotlin base types. 
- to make use of coroutines, since we were using them for room etc. 

We also looked into porting our code from volley to retrofit, as we wanted to use co-routines, and the kotlin retrofit client supports those (https://openapi-generator.tech/docs/generators/kotlin/) , but on looking at the code for those it was too far away from our manual volley code that we already had.

4brunu replied:
By the way, just curious, but why do you prefer Volley instead of OkHttp or Retrofit?

alisters replied:
I wouldn't say i have a preference really - i've never used the others. In kotlin i've mostly done server side stuff - using spring and web client recently. We are retrofitting a code gen/contract first approach to a set of existing clients. The android client already used volley. The author of that chose volley because the android doco seemed to default to using it, especially for kotlin based android apps. That said I like it - the design allowed mixing in various functionality easily using dependency injection with Dagger - so fairly painless. 

BTW if you haven't seen it from already the README does cover some of our design - and how we've tried to make the generator surface Volleys functionality. We don't use that much of it, but i tried to make the generator generic enough that other people could use and extend it for their use cases.

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/request/IRequestFactory.mustache:
```diff
@@ -0,0 +1,78 @@
+interface IRequestFactory {
+
+    companion object {
+        /**
+         * ISO 8601 date time format.
+         * @see https://en.wikipedia.org/wiki/ISO_8601
+         */
+        val DATE_TIME_FORMAT = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ")
```
@alisters We should use OffsetDateTime/LocalDate instead of Date and DateTimeFormatter instead of SimpleDateFormat everywhere. Or do you have a requirement to use Date?

alisters replied:
Changed to OffsetDateTime/LocalDate and DateTimeFormatter. Some additional points
- The formats used by the DateTimeFormatter are hardcoded. We lifted the previous code from a different open api library I think but don't use this code - in our api date's are represented as strings without formats. I believe this may have been due to incompatible code gen in different libraries that we use when a format was applied. 
- Petstore has a sample api with a string of format "Date Time" but not one of just "date"
TL;DR - i think the above code is unexercised and hence maybe untested, but agree with the change, and if this were done in all libraries maybe we could annotate our date times with formats.

4brunu replied:
According to the openapi spec, here are the date formats that should be used.

- `date` – full-date notation as defined by RFC 3339, section 5.6, for example, `2017-07-21`
- `date-time` – the date-time notation as defined by RFC 3339, section 5.6, for example, `2017-07-21T17:32:28Z`

https://swagger.io/docs/specification/data-models/data-types/

alisters replied:
In the modified code i use 
        fun formatDateTime(datetime: OffsetDateTime) = DateTimeFormatter.ISO_INSTANT.format(datetime)
        fun formatDate(date: LocalDate) = DateTimeFormatter.ISO_LOCAL_DATE.format(date)

I think i have these correct as https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html shows them as lining up with the swagger data-types/RFC references

4brunu reviewed (commented):
@alisters I left some comments on some things to improve.
I also saw some TODO's, are you planning to fix them before merging this, or is it future work?

alisters commented:
@4brunu - i'm addressing your comments and will add replies to each. Most look logical and just good practice - thank you 

With regards the TODO's - definitely future work. Many of them are around different forms of auth which we don't actually have a need to implement - so it's low down the priority list. 

I did raise a comment in the associated issue as to whether there was a feature matrix for generated clients or servers. If such a thing existed i'd like to surface what our library can and can't do, so that users of it know when they add something to an api spec as to whether or not the generated code will work. e.g We use other generators/libraries in some other parts of our code and they fail on anyOf constructs. I wonder if any of the "meta data" style flags that are present on the generated models could be used to generate compilation warnings ?

4brunu commented:
@alisters that matrix exists here, but I'm not sure if it's flexible enough to specify what features the Kotlin Client generator supports when using Volley.

https://github.com/OpenAPITools/openapi-generator/blob/b755ae288afe9b4d23de14e72ff6dc7735686690/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java#L122-L145

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/request/IRequestFactory.mustache:
```diff
@@ -0,0 +1,64 @@
+         * ISO 8601 date time format.
+         * @see https://en.wikipedia.org/wiki/ISO_8601
+         */
+        fun formatDateTime(datetime: OffsetDateTime) = DateTimeFormatter.ISO_INSTANT.format(datetime)
```
@alisters Here you should use DateTimeFormatter.ISO_OFFSET_DATE_TIME

alisters replied:
@4brunu - is there a reason for preferring ISO_OFFSET_DATE_TIME over ISO_INSTANT ? The link you supplied me https://swagger.io/docs/specification/data-models/data-types/, and it's relevant section 

date-time – the date-time notation as defined by RFC 3339, section 5.6, for example, **_2017-07-21T17:32:28Z_**  - would seem to suggest ISO_INSTANT ? 

I can't give you a solid reason for that, it's been a while since i had to think through this in our stack i just have a nagging feeling we landed on UTC representation for a reason. If you have a strong preference for ISO_OFFSET_DATE_TIME  i can try and dig up relevant design discussion notes. I don't think it will make a difference for this location, as mentioned we don't use the format for our date fields, i just think we might have chosen it to line up with how we serialize dates in other sections of our stack.

4brunu replied:
You are right, let's keep it as is. I suggested this because it's what the other kotlin generators use, but I think you are right. 👍

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/api.mustache:
```diff
@@ -0,0 +1,129 @@
+            {{#queryParams}}
+            "{{baseName}}" to IRequestFactory.parameterToString({{paramName}}),
+            {{/queryParams}}
+        ).filter { it.value != ""}
```
@alisters could you please replace `it.value != ""` with `it.value.isNotEmpty()`?

alisters replied:
Done

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/auth/apikeyauth.mustache:
```diff
@@ -0,0 +1,16 @@
+
+     val apiKeyAuthHeaderFactoryBuilder = {
+          paramName: String, apiKeyPrefix: String?, apiKey: String? -> {
+              mapOf("$paramName" to
```
@alisters could you please replace `mapOf("$paramName" to` with `mapOf(paramName to`?

alisters replied:
Fixed and this will resolve the issue below, when samples are regenerated

4brunu commented on samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/request/RequestFactory.kt:
```diff
@@ -0,0 +1,87 @@
+
+     val apiKeyAuthHeaderFactoryBuilder = {
+          paramName: String, apiKeyPrefix: String?, apiKey: String? -> {
+              mapOf("$paramName" to
```
@alisters could you please replace `mapOf("$paramName" to` with `mapOf(paramName to`?

alisters replied:
See above

4brunu replied:
Sorry, I didn't noticed that this was the generated code 👍

alisters commented:
@4brunu - i still need to regenerate the samples. That might not happen till tomorrow (Australia time)

4brunu commented:
Sure, take your time 👍

alisters commented:
@4brunu - samples partially regenerated. See comment above - https://github.com/OpenAPITools/openapi-generator/pull/10253#discussion_r768321651.

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,139 @@
+        mavenCentral()
+    }
+    dependencies {
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.5.10"
```
@alisters could you please reference the ext.kotlin_version here please?

alisters replied:
My bad, didn't pull the changes back into the mustache file. Fixed now

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/build.mustache:
```diff
@@ -0,0 +1,139 @@
+    }
+}
+
+ext {
+    kotlin_version = '1.5.10'
+    swagger_annotations_version = "1.6.2"
+    gson_version = "2.8.6"
+    volley_version = "1.2.0"
+    junit_version = "4.13.2"
+    robolectric_version = "4.5.1"
+    concurrent_unit_version = "0.4.6"
```
@alisters I think all of these versions are no longer needed, since they are also in the top of this file

alisters replied:
Fixed now

4brunu commented on samples/client/petstore/kotlin-jvm-volley/src/main/java/org/openapitools/client/infrastructure/Serializer.kt:
```diff
@@ -0,0 +1,22 @@
+import java.time.OffsetDateTime
+import java.util.UUID
+
+object Serializer {
```
@alisters I just noticed that this class is not used, which means that, OffsetDateTimeAdapter, LocalDateAdapter, ByteArrayAdapter are also not used.
Could you please remove the reference to them in the KotlinClientCodegen.java?
You should remove them from the `supportingFiles`

alisters replied:
The adapter classes are used inside GsonRequest - src/main/resources/kotlin-client/libraries/jvm-volley/request/GsonRequest.mustache

I've removed the Serialzer as a supporting file. In processJVMVolleyLibrary
  addSupportingSerializerAdapters(infrastructureFolder);
        supportingFiles.remove(new SupportingFile("jvm-common/infrastructure/Serializer.kt.mustache", infrastructureFolder, "Serializer.kt"));


The serializer class is part of jvm-common - src/main/resources/kotlin-client/jvm-common/infrastructure/Serializer.kt.mustache. So we left it in the supporting files that are generated when gson is used as the serializer, even though our library uses gson, but doesn't use it. I guess we avoided another conditional in kotlin code gen for volley AND gson - it's added addSupportingSerializerAdapters, which only knows about serializers at the moment. 

Let me know if https://github.com/OpenAPITools/openapi-generator/pull/10253/commits/d05a42b53eaf6f0348e8e23ccf8c09d69d9e1805  is ok

4brunu commented on modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java:
```diff
@@ -605,6 +702,11 @@ private void commonSupportingFiles() {
             supportingFiles.add(new SupportingFile("build.gradle.kts.mustache", "", "build.gradle.kts"));
             supportingFiles.add(new SupportingFile("settings.gradle.kts.mustache", "", "settings.gradle.kts"));
+        } else if (getLibrary().equals(JVM_VOLLEY)) {
+            supportingFiles.add(new SupportingFile("build.mustache", "", "build.gradle"));
+            supportingFiles.add(new SupportingFile("gradle.properties.mustache", "", "gradle.properties"));
+            supportingFiles.add(new SupportingFile("settings.gradle.mustache", "", "settings.gradle"));
+            supportingFiles.add(new SupportingFile("manifest.mustache", "", "src/main/AndroidManifest.xml"));
```
@alisters just wondering, how different are those grade files from the other shared gradle files?
If they are very different, we should keep them separated, if they are similar, we should consider merge them.

alisters replied:
The build.mustache/build.gradle is very different on account of android versus straight jvm/java builds. 
Same goes for the manifest - Android only. 
Other builds don't have a gradle properties - i could fold that into the build.gradle i think. Pretty new to gradle so i think myself or colleague Steve just made one by default - no strong rationale there.
The settings.gradle is the shared one - just keeping the conditional logic a bit cleaner there i guess ?

4brunu replied:
Looks good to me 👍

4brunu commented:
@alisters I have been investigating the question of have having one gson instance that is shared by all requests vs one gson instance per request, and I think we should go with the shared one, and use the `Serializer.kt`.
This way we could make the volley generator agnostic from the serializer, and support multiple ones.

So here is my sugestions:
Rename `GsonRequest` to something else, like `Request`.
In the GsonRequest.build() stop receiving the `type` parameter, and receive the response type as generic
`Serializer.gson.fromJson(bodyContent, (object: TypeToken<T>(){}).getType())`
`Serializer.gson.toJson(content, T::class.java)`

Basically the usages of gson should be only inside `GsonRequest`.

With this it will be very easy to all support for all the serializers, moshi, kotlinx serialization, etc.

4brunu commented on modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-volley/request/IRequestFactory.mustache:
```diff
@@ -0,0 +1,64 @@
+         * ISO 8601 date time format.
+         * @see https://en.wikipedia.org/wiki/ISO_8601
+         */
+        fun formatDateTime(datetime: OffsetDateTime) = DateTimeFormatter.ISO_INSTANT.format(datetime)
+        fun formatDate(date: LocalDate) = DateTimeFormatter.ISO_LOCAL_DATE.format(date)
```
@alisters here instead of using the DateFormatters directly, we should use the serializer to keep everything coherent

https://github.com/OpenAPITools/openapi-generator/blob/c27ddc67fe309298b2914c946ceb60b576c09a2b/samples/client/petstore/kotlin-gson/src/main/kotlin/org/openapitools/client/infrastructure/ApiClient.kt#L254-L262

https://github.com/OpenAPITools/openapi-generator/blob/c27ddc67fe309298b2914c946ceb60b576c09a2b/modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-okhttp/infrastructure/ApiClient.kt.mustache#L426-L450

alisters replied:
@4brunu - i have to think through this one. Would this still allow the application of request factory specific adapters if they are added to a single factory instance ?  i agree allowing the multiple serializers is a good move if it can happen.  

I'm on leave for 3 weeks from this afternoon - this work is part work time, part personal time - so i'm not sure when i'll get this done. Are you ok with this potentially restarting in the new year ? I imagine you don't mind if this goes in this release or the next point release ?

4brunu replied:
With the Serializer you can do `Serializer.gsonBuilder.add(...)`

What do you mean by test harness setup?

Sure, when you have time 👍 
Let's merge this when it's more solid, because now we can change whatever we want, once it's merge, we need to worry about backwards compatibility

4brunu replied:
@alisters since this PR was already merge, I would appreciate it if you could continue the changes we discuss here in another PR when you have time 🙂

alisters replied:
@4brunu - by test harness i just meant someway to run the generated code with the alternate serializers - thats all. I'll raise a new PR for this serializer related work.

wing328 commented:
As discussed with @4brunu , this PR is ready to be included in v5.3.1 so I've merged it.

For enhancements, please file separate PRs.

wing328 commented:
UPDATE: Included the tests in the CI via https://github.com/OpenAPITools/openapi-generator/pull/11156

alisters commented:
@wing328 and @4brunu  - thank you for merging this and @4brunu - thank you for all the refinements. I'll raise another PR for supporting additional serializers.
